captum/optim/_core/loss.py

```diff
@@ -237,28 +429,35 @@ def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
                     (activ_a - activ_b) ** 2
                 ).mean() / self.decay_ratio ** float(d)
 
-        return sum_tensor
+        return -sum_tensor
 
 
-class Direction(Loss):
+@loss_wrapper
+class Direction(BaseLoss):
     """
     Visualize a general direction vector.
     Carter, et al., "Activation Atlas", Distill, 2019.
     https://distill.pub/2019/activation-atlas/#Aggregating-Multiple-Images
     """
 
-    def __init__(self, target: nn.Module, vec: torch.Tensor) -> None:
-        super(Loss, self).__init__()
-        self.target = target
+    def __init__(
+        self,
+        target: nn.Module,
+        vec: torch.Tensor,
+        batch_index: Optional[int] = None,
+    ) -> None:
+        BaseLoss.__init__(self, target, batch_index)
         self.direction = vec.reshape((1, -1, 1, 1))
 
     def __call__(self, targets_to_values: ModuleOutputMapping) -> torch.Tensor:
         activations = targets_to_values[self.target]
         assert activations.size(1) == self.direction.size(1)
+        activations = activations[self.batch_index[0] : self.batch_index[1]]
         return torch.cosine_similarity(self.direction, activations)

```

Comment:
I'm not sure if we should be using `torch.cosine_similarity` here and for the other Direction objectives.

This is Lucid's cosine similarity code implemented in pytorch:

```
def dot_cossim(
    x: torch.Tensor,
    y: torch.Tensor,
    cossim_pow: float = 0.0,
    dim: int = 1,
    eps: float = 1.0e-4,
) -> torch.Tensor:
    """
    Compute the cosine similarity of two tensors.


    Args:
        x (torch.Tensor): The tensor that you wish to compute the cosine similarity
            for in relation to tensor y.
        y (torch.Tensor): The tensor that you wish to compute the cosine similarity
            for in relation to tensor x.
        cossim_pow (float, optional): The desired cosine similarity power to use.
        dim (int, optional): The target dimension for computing cosine similarity.
        eps (float, optional): If cossim_pow is greater than zero, the desired
            epsilon value to use for cosine similarity calculations.
    Return:
    
    """

    dot = torch.sum(x * y, dim)

    if cossim_pow == 0:
        return torch.mean(dot)

    x_mags = torch.sqrt(torch.sum(x * x, dim))
    y_mags = torch.sqrt(torch.sum(y * y, dim))

    cossims = dot / (eps + x_mags) / (eps + y_mags)
    floored_cossims = torch.max(torch.full_like(cossims, 0.1), cossims)
    return torch.mean(dot * floored_cossims ** cossim_pow)
```

https://github.com/tensorflow/lucid/blob/master/lucid/optvis/objectives_util.py#L15


I ran a few tests with the above code and compared it's results to `torch.cosine_similarity`. The results were significantly different results in some cases:


```

def pytorch_cossim(activ: torch.Tensor, vec: torch.Tensor, cossim_pow=0, eps=1.0e-4):
    if cossim_pow == 0:
        return torch.mean(torch.cosine_similarity(vec, activ, eps=eps))
    else:
        return torch.mean(
            torch.cosine_similarity(vec, activ, eps=eps, dim=1) ** cossim_pow
        )

test_activ = torch.randn(1, 2, 4, 4)
test_vec = torch.randn(1, 4)
cossim_power = 1.0 # Changing this number can make the results differ by a lot

print("Lucid cossim")
out = dot_cossim(test_activ, test_vec, cossim_power, dim=1)
print(out)

print("Captum cossim")
out = pytorch_cossim(test_activ, test_vec, cossim_power)
print(out)
```

Reply:
Ah thanks for the catch. Actually `torch.cosine_similarity` returns `cossims = dot / (eps + x_mags) / (eps + y_mags)` and not the `dot` part. (cosine similarity is dot product over product of magnitudes i.e. dot product of normalized vectors). I'll fix this and rename it, since cosine_similarity typically just means `cossims = dot / (eps + x_mags) / (eps + y_mags)`.

Reply:
Actually let's just use the implementation above. `dot_cossim` carries the connotation that we multiply dot product by cosine similarity.

Reply:
Thanks for the clarification on `torch.cosine_similarity`!